File: dp-app/app/internal/KnoxSso.scala
package internal

import java.io.ByteArrayInputStream

import scala.io.Source
import java.security.cert.CertificateFactory
import java.security.cert.Certificate
import java.security.PublicKey

import io.jsonwebtoken.{ExpiredJwtException, Jwts}
import java.nio.file.{Files, Paths}
import java.util.Date

import com.google.inject.Inject
import play.api.Configuration
import play.api.Logger

class KnoxSso @Inject()(configuration: Configuration) {
  private val DP_APP_HOME: String = configuration.underlying.getString("DP_APP_HOME")
  private val publicKeyPath = configuration.underlying.getString("signing.pub.key.path")
  private val signingKey: Option[PublicKey] = getSigningPublicKey
  private val ssoCookieName: String = configuration.underlying.getString("sso.cookie.name")
  private val knoxUrl: String = configuration.underlying.getString("knox.url")
  private val knoxWebssoPath: String = configuration.underlying.getString("knox.websso.path")

  def isSsoConfigured(): Boolean = {
    //TODO store this in db or elsewhere . user can setup sso or disable at will.
    return !DP_APP_HOME.isEmpty && Files.exists(Paths.get(pubFilePath))
  }

  def getSsoCookieName(): String = ssoCookieName

  def getLoginUrl(appRedirectUrl: String): String = s"${knoxUrl}${knoxWebssoPath}?originalUrl=${appRedirectUrl}"

  def validateJwt(serializedJWT: String): Either[Exception, (String, Long)] = {
    if (signingKey.isDefined) {
      try {
        val parsed = Jwts.parser().setSigningKey(signingKey.get).parseClaimsJws(serializedJWT)
        val expiration: Date = parsed.getBody.getExpiration
        Right((parsed.getBody.getSubject), if (expiration != null) expiration.getTime else -1)
      } catch {
        case e: ExpiredJwtException =>
          Left(new Exception("token-expired"))
      }
    } else {
      Logger.info("sso signig key not found")
      Left(new Exception("sso-not-configured"))
    }
  }

  private def pubFilePath: String = s"${publicKeyPath}"

  private def getSigningPublicKey(): Option[PublicKey] = {
    try {
      val source: Source = Source.fromFile(publicKeyPath)
      val certificateString: String = source.mkString
      val fact: CertificateFactory = CertificateFactory.getInstance("X.509")
      val is: ByteArrayInputStream = new ByteArrayInputStream(certificateString.getBytes("UTF8"))
      val cer: Certificate = fact.generateCertificate(is)
      Some(cer.getPublicKey)
    } catch {
      case e: Exception => {
        None
      }
    }
  }

}

File: services/gateway/src/main/java/com/hortonworks/dataplane/gateway/filters/KnoxSso.java
package com.hortonworks.dataplane.gateway.filters;

import com.google.common.base.Optional;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import org.apache.commons.io.FileUtils;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import java.io.ByteArrayInputStream;
import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.security.PublicKey;
import java.security.cert.CertificateException;
import java.security.cert.CertificateFactory;
import java.security.cert.X509Certificate;
import java.util.Date;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


@Component
public class KnoxSso {
  private static final Logger logger = LoggerFactory.getLogger(KnoxSso.class);

  @Value("${sso.cookie.name}:")
  private String ssoCookieName;

  @Value("${signing.pub.key.path:}")
  private String publicKeyPath;

  @Value("${knox.url:}")
  private String knoxUrl;

  @Value("${knox.websso.path:}")
  private String knoxWebssoPath;


  public  boolean isSsoConfigured(){
    if (publicKeyPath!=null && !publicKeyPath.isEmpty()) {
      boolean ssoConfigured = Files.exists(Paths.get(publicKeyPath));
      return ssoConfigured;
    }else{
      return false;
    }
  }

  public  String getSsoCookieName(){
    return ssoCookieName;
  }

  public String getLoginUrl(String appRedirectUrl){
    return String.format("%s%s?originalUrl=%s",knoxUrl,knoxWebssoPath,appRedirectUrl);
  }

  public Optional<String> validateJwt(String token){
    Claims claims =Jwts.parser()
      .setSigningKey(getSigningKey())
      .parseClaimsJws(token).getBody();
    Date expiration = claims.getExpiration();
    if (expiration.before(new Date())){
      Optional.absent();
    }
    return Optional.of(claims.getSubject());
  }

  private PublicKey getSigningKey(){
    try {
      String pemString = FileUtils.readFileToString(new File(publicKeyPath));
      CertificateFactory fact = CertificateFactory.getInstance("X.509");
      ByteArrayInputStream is = new ByteArrayInputStream(
        pemString.getBytes("UTF8"));
      X509Certificate cer = (X509Certificate) fact.generateCertificate(is);
      return cer.getPublicKey();
    } catch (IOException e) {
      logger.error("Exception",e);
      throw new RuntimeException(e);
    } catch (CertificateException e) {
      logger.error("Exception",e);
      throw new RuntimeException(e);
    }
  }
}